docs/cvl/invariants.md

```diff
@@ -75,6 +78,25 @@ and by {ref}`--parametric_contracts`. `View` and `pure` methods are excluded fro
 invariant checking as by definition they cannot change the state 
 of their contract. 
 
+A strong invariant performs the same checks as a weak invariant - i.e. it will be checked for the constructor
+and for any other method it will be assumed before executing the method (pre-state) and asserted afterward execution of the method (post-state). 
+In addition to these steps, a strong invariant also asserts and assumes the invariant _during_ method
+execution at location that potentially break the invariant. The invariant can be violated if there
+is an unresolved external call that is able to modify the state of the current contract. To verify the strong invariant, for every unresolved external call `c` 
+(a call that will force the prover to havoc storage), a strong invariant will do the following steps:
+
+ 1. Assert that the invariant holds before the `c` - if the invariant does not hold due to some logic of the current method, this will yield a counter example.
+ 2. Assume the invariant holds after the call `c`. The semantics is that the call did not break the invariant.  
+ 3. In the case `c` is a `delegatecall` havoc the current's contact storage and assert the invariant once more. 
```

Comment:
Not clear to me what exactly happens and why we do this. Does 3 replace step 2, or happens right after it?
Is there any realistic case where there is an invariant we can prove right after a havoc? (AFAIK resetting storage also resets ghosts)

Reply:
Step 3 is in addition and should provide soundness.  A delegatecall can change storage, so we simulate this by havocing storage.  We do this at the end, because we need to show that even after step 2, the delegator cannot break the invariant (we don't need it then after step 1, because the solver can always assume that step 2 does no change, so only checking after step 2 also catches the bug).

I think step3 does not havoc ghost (is this right?)  The idea is that the delegator cannot change the ghost state or the state of any other contract, just the contract that it is called on.  

This is mostly for invariants that don't even talk about the state of the contract. If the contract that does the delegate call is not the main contract we want to reason about, then delegatecalls from this contract are always okay.

Reply:
> I think step3 does not havoc ghost (is this right?)

Correct. 

I updated the description to be more meaningful. Is it better now?